### ISSUE TITLE
How to discuss latency...

### ISSUE THREAD

**Issue by fiestajetsam** [moq-wg/moq-requirements#72]
Based on Cullen's feedback, we should have a better way to talk about latency with the use cases and how they differ, ideally avoiding units of measure but without being too vague. There should also be clarity over factoring variances in network topology too.

**Comment by huitema** [comment]
There are at least two ways to address that question. We could discuss the relation between human perception and end-to-end signals, a.k.a., psychophysics. Or, we could have soft goals for how much of the end-to-end latency budget goes to what component of the transmission chain. Both approaches have problems, so we may also chose to say nothing.

Usability studies (and psychophysics) have produced known results for the effect of latency on communication. For example, we know that for virtual reality, the movements of a user's head need to be reflected on the virtual display very quickly, in a few milliseconds. If you don't get that, the VR is less convincing, and we can even see ill effects like inducing nausea. There are similar issues with some real time games, such as for example fencing, which are only fun to play if the delays are less than 15ms. And of course there are the well known results about spontaneous turn taking in audio communication, which only works well if the delays are less than 150ms. These numbers are useful when designing systems, but they are less useful when evaluating protocols. After all, protocols can never beat the speed of light, and we will still want to communicate somehow even if participants are so far away that we cannot bring the latency under 500ms or even a few seconds. 

We could also have goals such as "not designing protocols that increase latency too much". For example, we may say that 99% of the time, the end-to-end latency shall be lower than 2RTT, allowing for one round of error correction. But such goals are a bit squishy. If the RTT is less than 2ms, should the protocol really aim for less than 4 ms? If the packet loss rate is 20%, is it better to cap the delay at 2 RTT and suffer 4% residual loss, or should the system allow for longer delay and fewer losses? Or should it go for higher compression and devote more bandwidth to FEC? We may have ideas about all that, but different systems will most likely make different tradeoffs.

Bottom line, it is much probably better to not have specific delay constraints in a requirement document.

**Comment by wilaw** [comment]
I concur with avoiding units of measure and specific thresholds. These will obsolete themselves very quickly based on technology evolution and application. 

I did like the intentionally boarder language that Luke introduced during his presentation at IEFT#115. He broke latency in to 3 broad buckets based on the application space: 

- **Real-time communications**: latency sufficiently low for real-time, bi-directional audio and video communications and game play
- **Interactive**: latency sufficiently low to enable interaction with a live stream, via text, emoji, or other non real-time user action.
- **Non-interactive**:  latency high enough that it allows only non-interactive consumption. 

The boundaries for these will vary with network protocol, media protocol, topology, loss, RTT, etc which is fine. We can't possible define all permutations of these within the requirements doc and nor should we.

**Comment by mortenvp** [comment]
As @huitema mentions, it can be helpful to provide examples of different latency thresholds required by various applications. However, I also agree that designing toward specific thresholds is probably less meaningful. As @wilaw also mentions, they will quickly be obsolete as new applications emerge. 

In my view, the application should be able to provide a latency budget to the transport layer. I find it difficult to see how one can build a "low-latency" protocol without the application providing what "low-latency" means. There will be cases where the latency goals cannot be met - in such cases, the transport should inform the application.

**Comment by VMatrix1900** [comment]
> In my view, the application should be able to provide a latency budget to the transport layer. I find it difficult to see how one can build a "low-latency" protocol without the application providing what "low-latency" means. There will be cases where the latency goals cannot be met - in such cases, the transport should inform the application.
That is exactly what [DTP](https://datatracker.ietf.org/doc/draft-shi-quic-dtp/) is proposing. Do you think it is useful to MoQ?

**Comment by mortenvp** [comment]
@VMatrix1900 Yes, the overall idea is there. I'm not sure about the implementation, though. E.g., is a QUIC extension needed - or can we layer this on top of the existing QUIC streams or datagrams? Also, the "block" approach seems to require the full block to be available at the sender before packetization and transmission. I'm not sure that is the best approach for some of the more "streaming" like applications.

**Comment by VMatrix1900** [comment]
>  I'm not sure about the implementation, though. E.g., is a QUIC extension needed - or can we layer this on top of the existing QUIC streams or datagrams? 

Good question. There are two parts of DTP: scheduling and FEC. For scheduling to work on top of QUIC. QUIC need to let the application set the priority of each stream dynamically. For FEC, since it is operating based on the packet, it needs the access to the packetization process so it can only work on top of QUIC datagram.

> Also, the "block" approach seems to require the full block to be available at the sender before packetization and transmission. I'm not sure that is the best approach for some of the more "streaming" like applications.

For the media streaming application, the "stream" is actually a series of blocks of data. For example, video stream is composed of a series of frame/GOP which is a basic processing unit of encoder/decoder.

**Comment by mortenvp** [comment]
> Good question. There are two parts of DTP: scheduling and FEC. For scheduling to work on top of QUIC. QUIC need to let the application set the priority of each stream dynamically. For FEC, since it is operating based on the packet, it needs the access to the packetization process so it can only work on top of QUIC datagram.

I believe you are correct. QUIC datagram would be needed to combine this mechanism with FEC. 

One of the reasons I would like to see the application specify the deadline is that I think the transport layer should combine this knowledge with information about the network. I.e. given the network latency does it make sense to (re-)transmit this data? Unless network latency information is exposed to the application such that it can cancel streams I think this should be done by the transport layer. 

Hope it makes sense?

> For the media streaming application, the "stream" is actually a series of blocks of data. For example, video stream is composed of a series of frame/GOP which is a basic processing unit of encoder/decoder.

In this case you are right - however sometimes e.g. if implementing a "relay" you don't really know how many e.g. MPEG-TS UDP packets constitute a frame. So how long do you wait? 

Hope that also makes sense?

**Comment by VMatrix1900** [comment]
> One of the reasons I would like to see the application specify the deadline is that I think the transport layer should combine this knowledge with information about the network. I.e. given the network latency does it make sense to (re-)transmit this data? Unless network latency information is exposed to the application such that it can cancel streams I think this should be done by the transport layer.
> 
> Hope it makes sense?

Exactly. The transport layer is closer to raw network information than application layer which gives it natural advantage to implement such kind of advanced scheduling.


> In this case you are right - however sometimes e.g. if implementing a "relay" you don't really know how many e.g. MPEG-TS UDP packets constitute a frame. So how long do you wait?

That is why a BLOCK_INFO frame is defined to convey the metadata of the block such as the size and deadline. The other end can optimize based on these information. For relay to works under the deadline API, the similar information needs to be passed between the chain of relays.

**Comment by mortenvp** [comment]
> That is why a BLOCK_INFO frame is defined to convey the metadata of the block such as the size and deadline. The other end can optimize based on these information. For relay to works under the deadline API, the similar information needs to be passed between the chain of relays.

I think I understand. However, the issue is if you consume, e.g., a MPEG-TS stream via a UDP socket. You don't have a BLOCK_INFO frame containing any metadata about the stream. You get UDP packets, and you need to figure out what constitutes a frame/block. This may not be so easy.

**Comment by VMatrix1900** [comment]
There are two ways to solve it:
1. the relay act as the QUIC endpoint and decode the MoQ stream(metadata + media data). The BLOCK_INFO frame proposed in [DTP](https://datatracker.ietf.org/doc/draft-shi-quic-dtp/) belongs to this approach. 
2. Prepend the metadata in every packet, like [MOQ Extensions](https://datatracker.ietf.org/doc/draft-defoy-moq-relay-network-handling/) is doing.

**Comment by mortenvp** [comment]
Hi @VMatrix1900 I'm not sure I understand how those would solve it (at least #1 I have to investigate #2). I have a legacy video encoder - it produces a UDP flow (I can look into the UDP packets and decode them - but I would rather not). I don't know which UDP packets belong to which video frame, they trickle in one by one (although in a burst).

**Comment by SpencerDawkins** [comment]
@fiestajetsam and @SpencerDawkins discussed this issue. 

We agree with many of the points people have made in comments here. 

We think we've addressed many of them in [Distinguishing between Interactive and Live Streaming Use Cases](https://moq-wg.github.io/moq-requirements/draft-ietf-moq-requirements.html#name-distinguishing-between-inte) in the current draft. 

What we haven't talked about in the draft, so far, is anything like deadline scheduling/delivery deadlines that have been proposed in individual drafts, but the drafts have not received enough discussion to be either accepted (as ideas) or rejected. 

We can leave this issue open, and see if these ideas do pop up again, whether on mailing lists, on agendas, in GitHub, or in Slack.

**Comment by SpencerDawkins** [comment]
This issue should have been closed by [PR 95](https://github.com/moq-wg/moq-requirements/pull/95), before IETF 117. I'll close it now.
